core/src/test/java/com/datastax/oss/driver/internal/core/CassandraProtocolVersionRegistryHighestCommonTest.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2017-2017 DataStax Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core;
+
+import com.datastax.oss.driver.api.core.CassandraVersion;
+import com.datastax.oss.driver.api.core.CoreProtocolVersion;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.google.common.collect.ImmutableList;
+import java.util.Collection;
+import java.util.Collections;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import static com.datastax.oss.driver.Assertions.assertThat;
+
+/**
+ * Covers {@link CassandraProtocolVersionRegistry#highestCommon(Collection)} separately, because it
+ * relies explicitly on {@link CoreProtocolVersion} as the version implementation.
+ */
+public class CassandraProtocolVersionRegistryHighestCommonTest {
+
+  private CassandraProtocolVersionRegistry registry = new CassandraProtocolVersionRegistry("test");
+
+  @Test
+  public void should_pick_v3_when_at_least_one_node_is_2_1() {
+    assertThat(
+            registry.highestCommon(
+                ImmutableList.of(mockNode("2.2.1"), mockNode("2.1.0"), mockNode("3.1.9"))))
+        .isEqualTo(CoreProtocolVersion.V3);
+  }
+
+  @Test
+  public void should_pick_v4_when_all_nodes_are_2_2_or_more() {
+    assertThat(
+            registry.highestCommon(
+                ImmutableList.of(mockNode("2.2.0"), mockNode("2.2.1"), mockNode("3.1.9"))))
+        .isEqualTo(CoreProtocolVersion.V4);
+  }
+
+  @Test
+  public void should_treat_rcs_as_next_stable_versions() {
+    assertThat(
+            registry.highestCommon(
+                ImmutableList.of(mockNode("2.2.1"), mockNode("2.1.0-rc1"), mockNode("3.1.9"))))
+        .isEqualTo(CoreProtocolVersion.V3);
+    assertThat(
+            registry.highestCommon(
+                ImmutableList.of(mockNode("2.2.0-rc2"), mockNode("2.2.1"), mockNode("3.1.9"))))
+        .isEqualTo(CoreProtocolVersion.V4);
+  }
+
+  @Test
+  public void should_skip_nodes_that_report_null_version() {
+    assertThat(
```

Comment:
Good idea to have a test for this,  the system.peers table has been unpredictable in the past and this is definitely possible.